src/libasr/codegen/x86_assembler.cpp

```diff
@@ -283,94 +329,149 @@ void emit_print_float(X86Assembler &a, const std::string &name) {
+    for (size_t i = 0; i < padding_size; i++) {
+        code.push_back(al, 0);
+    }
+ }
 
-void emit_elf64_footer(X86Assembler &a) {
-    a.add_var("ehdrsize", "ehdr", "phdr");
-    a.add_var("phdrsize", "phdr", "text_phdr");
-    a.add_var64("e_phoff", "ehdr", "phdr");
-    a.add_var64("phdr_size", "ehdr", "text_segment_start");
-    a.add_var64("text_segment_offset", "ehdr", "text_segment_start");
-    a.add_var64("text_segment_size", "text_segment_start", "text_segment_end");
-    a.add_var64("data_segment_offset", "ehdr", "data_segment_start");
-    a.add_var64("data_segment_size", "data_segment_start", "data_segment_end");
-}
+Vec<uint8_t> create_elf64_x86_binary(Allocator &al, X86Assembler &a) {
 
-void emit_exit_64(X86Assembler &a, std::string name, int exit_code) {
-    a.add_label(name);
-    // void exit(int status);
-    a.asm_mov_r64_imm64(LCompilers::X64Reg::rax, 60); // sys_exit
-    a.asm_mov_r64_imm64(LCompilers::X64Reg::rdi, exit_code); // exit code
-    a.asm_syscall(); // syscall
+    /*
+        The header segment is a segment which holds the elf and program headers.
+        Its size currently is
+            sizeof(Elf64_Ehdr) + 3 * sizeof(Elf64_Phdr)
+            that is, 64 + 3 * 56 = 232
+        Since, it is a segment, it needs to be aligned by boundary 0x1000
+        (we add temporary zero bytes as padding to accomplish this alignment)
+
+        Thus, the header segment size for us currently is 0x1000.
+
+        For now, we are hardcoding this size here.
+
+        TODO: Later compute this header segment size dynamically depending
+        on the different segments present
+    */
+    const int HEADER_SEGMENT_SIZE = 0x1000;
+
+    // adjust/offset the origin address as per the extra bytes of HEADER_SEGMENT_SIZE
+    uint64_t origin_addr = a.origin() - HEADER_SEGMENT_SIZE;
+
+    Elf64_Ehdr e = get_elf_header(a.get_defined_symbol("_start").value);
+    Elf64_Phdr p_program = get_seg_header(4, origin_addr, HEADER_SEGMENT_SIZE, 0, 0);
+
+    const uint64_t text_seg_size = a.get_defined_symbol("text_segment_end").value - a.get_defined_symbol("text_segment_start").value;
+    Elf64_Phdr p_text_seg = get_seg_header(5, origin_addr, text_seg_size, p_program.offset, p_program.filesz);
+
+    const uint64_t data_seg_size = a.get_defined_symbol("data_segment_end").value - a.get_defined_symbol("data_segment_start").value;
+    Elf64_Phdr p_data_seg = get_seg_header(6, origin_addr, data_seg_size, p_text_seg.offset, p_text_seg.filesz);
+
+    Vec<uint8_t> bin;
+    bin.reserve(al, HEADER_SEGMENT_SIZE);
+
+    {
+        append_header_bytes(al, e, bin);
+        append_header_bytes(al, p_program, bin);
+        append_header_bytes(al, p_text_seg, bin);
+        append_header_bytes(al, p_data_seg, bin);
+
+        LCompilers::align_by_byte(al, bin, 0x1000);
+    }
+
+    for (auto b:a.get_machine_code()) {
+        bin.push_back(al, b);
+    }
```

Comment:
Is this copying the (potentially big) binary code byte by byte?

Reply:
Yes, it is copying the binary code byte by byte. Another approach that we can do is to just return the ELF and Phdr headers in `create_elf64_x86_binary()` (we can rename it as `create_elf64_x86_header()`).

For example:
```bash
void X86Assembler::save_binary64(const std::string &filename) {
    Vec<uint8_t> header = create_elf64_x86_header(m_al, *this);
    {
        std::ofstream out;
        out.open(filename);
        out.write((const char*) header.p, header.size());
        out.write((const char*) m_code.p, m_code.size());
        out.close();
    }
#ifdef LFORTRAN_LINUX
    std::string mode = "0755";
    int mod = strtol(mode.c_str(), 0, 8);
    if (chmod(filename.c_str(),mod) < 0) {
        throw AssemblerError("chmod failed");
    }
#endif
}
```

Reply:
I currently updated it to return the `header` only.